File: benchmarks/C++/src/harness.cpp
#include <iostream>
#include <string>

#include "run.h"

static Run process_arguments(int32_t argc, const char* argv[]) {  // NOLINT
  std::string name(argv[1]);
  Run run(name);

  if (argc > 2) {
    run.set_num_iterations(std::stoi(argv[2]));
    if (argc > 3) {
      run.set_inner_iterations(std::stoi(argv[3]));
    }
  }

  return run;
}

static void print_usage() {
  std::cout << "./harness benchmark [num-iterations [inner-iter]]\n";
  std::cout << "\n";
  std::cout << "  benchmark      - benchmark class name\n";
  std::cout << "  num-iterations - number of times to execute benchmark, "
               "default: 1\n";
  std::cout << "  inner-iter     - number of times the benchmark is executed "
               "in an inner loop,\n";
  std::cout << "                   which is measured in total, default: 1\n";
}

int main(int argc, const char* argv[]) {
  if (argc < 2) {
    print_usage();
    return 1;
  }

  Run run = process_arguments(argc, argv);
  run.run_benchmark();
  run.print_total();

  return 0;
}

File: benchmarks/Rust/src/bin/harness.rs
use awfy::run::Run;
use std::env::args;

fn process_arguments() -> Run {
    let mut args = args();
    args.next(); // Skip binary name
    let mut run = Run::new(args.next().unwrap());
    if let Some(arg) = args.next() {
        run.set_num_iterations(arg.parse().unwrap());
        if let Some(arg) = args.next() {
            run.set_inner_iterations(arg.parse().unwrap());
        }
    }
    run
}

fn print_usage() {
    println!("Harness [benchmark] [num-iterations [inner-iter]]");
    println!();
    println!("  benchmark      - benchmark class name ");
    println!("  num-iterations - number of times to execute benchmark, default: 1");
    println!("  inner-iter     - number of times the benchmark is executed in an inner loop, ");
    println!("                   which is measured in total, default: 1");
}

fn main() {
    if args().len() < 2 {
        print_usage();
        return;
    }

    let mut run = process_arguments();
    run.run_benchmark();
    run.print_total();
}
